code/datums/components/_component.dm

```diff
@@ -411,74 +439,95 @@
  *
  * Properly handles duplicate situations based on the `dupe_mode` var
  */
-/datum/proc/_AddComponent(list/raw_args)
-	var/new_type = raw_args[1]
-	var/datum/component/nt = new_type
+/datum/proc/_AddComponent(list/raw_args, source)
+	var/original_type = raw_args[1]
+	var/datum/component/component_type = original_type
 
 	if(QDELING(src))
-		CRASH("Attempted to add a new component of type \[[nt]\] to a qdeleting parent of type \[[type]\]!")
+		CRASH("Attempted to add a new component of type \[[component_type]\] to a qdeleting parent of type \[[type]\]!")
 
-	var/dm = initial(nt.dupe_mode)
-	var/dt = initial(nt.dupe_type)
+	var/dupe_mode = initial(component_type.dupe_mode)
+	var/dupe_type = initial(component_type.dupe_type)
+	var/uses_sources = initial(component_type.uses_sources)
+	if(uses_sources && !source)
+		CRASH("Attempted to add a sourced component of type '[component_type]' to '[type]' without a source!")
 
-	var/datum/component/old_comp
-	var/datum/component/new_comp
+	var/datum/component/old_component
+	var/datum/component/new_component
 
-	if(ispath(nt))
-		if(nt == /datum/component)
-			CRASH("[nt] attempted instantiation!")
+	if(ispath(component_type))
+		if(component_type == /datum/component)
+			CRASH("[component_type] attempted instantiation!")
 	else
-		new_comp = nt
-		nt = new_comp.type
+		new_component = component_type
+		component_type = new_component.type
 
 	raw_args[1] = src
-
-	if(dm != COMPONENT_DUPE_ALLOWED && dm != COMPONENT_DUPE_SELECTIVE)
-		if(!dt)
-			old_comp = GetExactComponent(nt)
+	if(dupe_mode != COMPONENT_DUPE_ALLOWED && dupe_mode != COMPONENT_DUPE_SELECTIVE)
+		if(!dupe_type)
+			old_component = GetExactComponent(component_type)
 		else
-			old_comp = GetComponent(dt)
-		if(old_comp)
-			switch(dm)
+			old_component = GetComponent(dupe_type)
+
+		if(old_component)
+			switch(dupe_mode)
 				if(COMPONENT_DUPE_UNIQUE)
-					if(!new_comp)
-						new_comp = new nt(raw_args)
-					if(!QDELETED(new_comp))
-						old_comp.InheritComponent(new_comp, TRUE)
-						QDEL_NULL(new_comp)
+					if(!new_component)
+						new_component = new component_type(raw_args)
+					if(!QDELETED(new_component))
+						old_component.InheritComponent(new_component, TRUE)
+						QDEL_NULL(new_component)
+
 				if(COMPONENT_DUPE_HIGHLANDER)
```

Comment:
I'm killing you

Reply:
Plz explain

Reply:
assuming components should have sources in the first place, I don't think they all should. seems like something that ought to be a dupe mode to me, since yaknow, that's what it does

Reply:
and so shall it be done